### PR TITLE
Signed and notarized automated macOS builds

### PR DESCRIPTION
**Pull request by jbmorley**
This change goes some way to automating the full release process by updating the build script to automatically sign and notarise the macOS builds. While it makes use of Fastlane's `notarize` and `match` actions, it doesn't migrate the whole build process to Fastlane (I tried that) because it just seems to add unnecessary complexity as Fastlane's trying to do too much magic under the hood and guesses wrong.

Longer-term I'd like to add support for using some flavour of semantic versioning and conventional commits to auto-generate the release notes and version number.

### CODE REVIEW

File: fastlane/Fastfile [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    )
+    zip(
+        path: target,
+        output_path: zip_path
+    )
+
+end
+
+lane :test do
+    sh("../scripts/test.sh")
```
Comment (pvinis) [inline fastlane/Fastfile:56]: love it

File: macos/Debug.xcconfig [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+// Configuration settings file format documentation can be found at:
+// https://help.apple.com/xcode/#/dev745c5c974
+
+#include "Common.xcconfig"
```
Comment (pvinis) [inline macos/Debug.xcconfig:24]: and love it

Comment (jbmorley) [reply macos/Debug.xcconfig:21]: Now I just need to move some of the settings into this!

Comment (jbmorley) [reply macos/Debug.xcconfig:21]: See https://github.com/jbmorley/fileaway/issues/90.

File: fastlane/Fastfile [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        git_url: ENV['CERTIFICATE_REPOSITORY'],
+        git_basic_authorization: ENV['CERTIFICATE_REPOSITORY_AUTHORIZATION_KEY'],
+        type: "developer_id",
+        app_identifier: "uk.co.inseven.fileaway",
+        readonly: true,
+        skip_provisioning_profiles: true,
+        keychain_name: keychain_path,
+        keychain_password: ENV['TEMPORARY_KEYCHAIN_PASSWORD'],
+    )
+end
```
Comment (jbmorley) [inline fastlane/Fastfile:32]: This change should introduce documentation explaining how to import a Mac development certificate into the match certificate store given how broken it is.

File: fastlane/Fastfile [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+#
+
+skip_docs
+
+keychain_path = Pathname.new(File.join(Dir.pwd, "..", "temp/temporary.keychain")).cleanpath.to_s
+
+lane :import_certificates do
+    # In theory, this will create the keys and download them if necessary.
+    # Failing because it somehow maps a Developer ID Application key to AppStore by mistake.
+    # We do this in a separate lane to ensure it doesn't taint our build (which is working fine, thank you very much).
```
Comment (jbmorley) [inline fastlane/Fastfile:20]: Clean up this documentation.

File: scripts/test.sh [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+# Determine the version and build number.
+# TODO: Get Xcode to synthesise these itself so that builds also work there.
+VERSION_NUMBER=`cat macos/version.txt | tr -d '[:space:]'`
+GIT_COMMIT=`git rev-parse --short HEAD`
+TIMESTAMP=`date +%s`
+BUILD_NUMBER="${GIT_COMMIT}.${TIMESTAMP}"
+
+# Import the certificates into our dedicated keychain.
+fastlane import_certificates
```
Comment (jbmorley) [inline scripts/test.sh:79]: Fastlane currently complains that it's being run without Bundler. It wants us to call:

```bash
 bundle exec fastlane import_certificates
```

File: scripts/test.sh [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+# Show the code signing details.
+codesign -dvv "$BUILD_DIRECTORY/Fileaway.app"
+
+# Notarize the release build.
+fastlane notarize_release
+
+# Archive the results.
+pushd "$BUILD_DIRECTORY"
+zip -r "Fileaway-macOS-${VERSION_NUMBER}-${BUILD_NUMBER}.zip" "."
+popd
```
Comment (jbmorley) [inline scripts/test.sh:97]: This script should almost certainly clean up the temporary keychain once it's done.

File: scripts/test.sh [hunk 7]
```diff
@@ -0,0 +1,97 @@
+#!/bin/bash
```
Comment (jbmorley) [inline scripts/test.sh:1]: This should totally be renamed to build.sh again.

File: scripts/test.sh [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+# Create the a new keychain.
+if [ -d "$TEMPORARY_DIRECTORY" ] ; then
+    rm -rf "$TEMPORARY_DIRECTORY"
+fi
+mkdir -p "$TEMPORARY_DIRECTORY"
+security create-keychain -p "$TEMPORARY_KEYCHAIN_PASSWORD" "$KEYCHAIN_PATH"
+security set-keychain-settings -lut 21600 "$KEYCHAIN_PATH"
+
+# Determine the version and build number.
+# TODO: Get Xcode to synthesise these itself so that builds also work there.
```
Comment (jbmorley) [inline scripts/test.sh:72]: Yes. We should definitely be doing this.

File: .github/workflows/build.yaml [hunk 9]
```diff
@@ -25,5 +25,18 @@ jobs:
       run: |
         for ip in $(dig @8.8.8.8 github.com +short); do ssh-keyscan github.com,$ip; ssh-keyscan $ip; done 2>/dev/null >> ~/.ssh/known_hosts
 
-    - name: Build
-      run: exec ./scripts/build.sh
+    - name: Build and test
+      env:
```
Comment (jbmorley) [inline .github/workflows/build.yaml:29]: Document these environment variables.

File: fastlane/Fastfile [hunk 10]
```diff
@@ -0,0 +1,28 @@
+skip_docs
+
+keychain_path = Pathname.new(File.join(Dir.pwd, "..", "temp/temporary.keychain")).cleanpath.to_s
```
Comment (jbmorley) [inline fastlane/Fastfile:3]: Pass this in as a parameter--there should be no magic in this file.

Comment (jbmorley) [reply fastlane/Fastfile:3]: Done.

File: scripts/test.sh (hunk 7 above)
Comment (jbmorley) [reply scripts/test.sh:1]: The build script should also have a copyright notice.

Comment (jbmorley) [reply scripts/test.sh]: Done.

File: scripts/build.sh [hunk 11]
```diff
@@ -3,17 +3,30 @@
 set -e
 set -o pipefail
 set -x
+set -u
```
Comment (jbmorley) [inline scripts/build.sh:6]: Add a copyright notice to the top of this file.

File: fastlane/Fastfile [hunk 12]
```diff
@@ -0,0 +1,24 @@
+skip_docs
```
Comment (jbmorley) [inline fastlane/Fastfile:1]: Add a copyright notice to the top of this file.

File: scripts/test.sh (hunk 5 above)
Comment (jbmorley) [reply scripts/test.sh]: Decided to reduce complexity by dropping Bundler for the time being. Long-term I'd love to remove the dependency on Fastlane, so hopefully we'll get away with this.

File: scripts/test.sh (hunk 6 above)
Comment (jbmorley) [reply scripts/test.sh]: Done.

File: scripts/test.sh (hunk 8 above)
Comment (jbmorley) [reply scripts/test.sh]: Added a TODO for this.

File: scripts/build.sh (hunk 11 above)
Comment (jbmorley) [reply scripts/build.sh:26]: Done.

File: fastlane/Fastfile (hunk 12 above)
Comment (jbmorley) [reply fastlane/Fastfile:21]: Done.

File: fastlane/Fastfile (hunk 3 above)
Comment (jbmorley) [reply fastlane/Fastfile:36]: Done.

**Pull request merged**